src/NuGet.Clients/PackageManagement.UI/Actions/UIActionEngine.cs

```diff
@@ -38,28 +38,142 @@ public UIActionEngine(ISourceRepositoryProvider sourceProvider, NuGetPackageMana
             {
-                uiService.ShowProgressDialog(windowOwner);
+                var installedPackages = await project.GetInstalledPackagesAsync(token);
+                HashSet<string> packageIds = new HashSet<string>(StringComparer.OrdinalIgnoreCase);
+                foreach (var p in installedPackages)
+                {
+                    packageIds.Add(p.PackageIdentity.Id);
+                }
+                var packagesToUpdateInProject = packagesToUpdate.Where(
+                    package => packageIds.Contains(package.Id)).ToList();
 
-                var projects = uiService.Projects;
+                if (packagesToUpdateInProject.Any())
+                {
+                    var includePrerelease = packagesToUpdateInProject.Where(
+                        package => package.Version.IsPrerelease).Any();
+
+                    var resolutionContext = new ResolutionContext(
+                        uiService.DependencyBehavior,
+                        includePrelease: includePrerelease,
+                        includeUnlisted: false,
+                        versionConstraints: VersionConstraints.None);
+                    var sources = new SourceRepository[] { uiService.ActiveSource };
+                    var actions = await _packageManager.PreviewUpdatePackagesAsync(
+                        packagesToUpdateInProject,
+                        project,
+                        resolutionContext,
+                        uiService.ProgressWindow,
+                        sources,
+                        sources,
+                        token);
+                    resolvedActions.AddRange(actions.Select(action => Tuple.Create(project, action))
+                        .ToList());
+                }
+            }
 
-                // TODO: should stable packages allow prerelease dependencies if include prerelease was checked?
-                // Allow prerelease packages only if the target is prerelease
-                var includePrelease = userAction.PackageIdentity.Version.IsPrerelease || userAction.Action == NuGetProjectActionType.Uninstall;
-                var includeUnlisted = userAction.Action == NuGetProjectActionType.Uninstall;
+            return resolvedActions;
+        }
 
-                var resolutionContext = new ResolutionContext(uiService.DependencyBehavior, includePrelease, includeUnlisted, VersionConstraints.None);
+        /// <summary>
+        /// The internal implementation to perform user action.
+        /// </summary>
+        /// <param name="resolveActionsTask">A function that returns a task that resolves the user 
+        /// action into project actions.</param>
+        /// <param name="executeActionsTask">A function that returns a task that executes 
+        /// the project actions.</param>
+        private async Task PerformActionImplAsync(
+            INuGetUI uiService,
+            Func<Task<IReadOnlyList<Tuple<NuGetProject, NuGetProjectAction>>>> resolveActionsTask,
```

Comment:
Extracted the common code into this new method. Then it's called by both PerformActionAsync() and PerformUpdateAsync().